src/WebJobs.Extensions.DurableTask.Analyzers/Analyzers/ActivityFunction/FunctionReturnTypeAnalyzer.cs

```diff
@@ -31,33 +31,30 @@ public class FunctionReturnTypeAnalyzer
                 var definition = functionDefinitions.Where(x => x.FunctionName == invocation.FunctionName).FirstOrDefault();
                 if (definition != null && invocation.ReturnTypeNode != null)
                 {
-                    if (TryGetInvocationReturnType(semanticModel, invocation, out ITypeSymbol invocationReturnType)
-                        && TryGetDefinitionReturnType(semanticModel, definition, out ITypeSymbol definitionReturnType))
+                    if (!IsValidReturnTypeForDefinition(invocation, definition))
                     {
-                        if (!IsValidReturnTypeForDefinition(invocationReturnType, definitionReturnType))
-                        {
-                            var diagnostic = Diagnostic.Create(Rule, invocation.InvocationExpression.GetLocation(), invocation.FunctionName, definitionReturnType.ToString(), invocationReturnType.ToString());
+                        var diagnostic = Diagnostic.Create(Rule, invocation.InvocationExpression.GetLocation(), invocation.FunctionName, definition.ReturnType, invocation.ReturnType);
 
-                            context.ReportDiagnostic(diagnostic);
-                        }
+                        context.ReportDiagnostic(diagnostic);
                     }
                 }
             }
         }
 
-        private static bool IsValidReturnTypeForDefinition(ITypeSymbol invocationReturnType, ITypeSymbol definitionReturnType)
+        private static bool IsValidReturnTypeForDefinition(ActivityFunctionCall invocation, ActivityFunctionDefinition definition)
         {
+            var definitionReturnType = definition.ReturnType;
             if (TryGetTaskTypeArgument(definitionReturnType, out ITypeSymbol taskTypeArgument))
             {
                 definitionReturnType = taskTypeArgument;
             }
 
-            return SyntaxNodeUtils.IsMatchingDerivedOrCompatibleType(definitionReturnType, invocationReturnType);
+            return SyntaxNodeUtils.IsMatchingDerivedOrCompatibleType(definitionReturnType, invocation.ReturnType);
         }
 
         private static bool TryGetTaskTypeArgument(ITypeSymbol returnType, out ITypeSymbol taskTypeArgument)
         {
-            if (returnType.Name.Equals("Task") && returnType is INamedTypeSymbol namedType)
+            if (returnType != null && returnType.Name.Equals("Task") && returnType is INamedTypeSymbol namedType)
             {
```

Comment:
small nit: Could simplify to `if (returnType is INameTypSymbol namedType && namedType.Name.Equals("Task"))`.  You can do this because `is` does an implicit check for `null`.